### PR TITLE
add test case for user

### PR DESCRIPTION
**Pull request by wemeya**

### CODE REVIEW

File: api/harborapi_test.go [hunk 1]
```diff
@@ -77,19 +80,28 @@ func init() {
 
 }
 
-func request(_sling *sling.Sling, acceptHeader string, authInfo ...usrInfo) (int, []byte, error) {
+func request(_sling *sling.Sling, acceptHeader string, authInfo ...usrInfo) (int, int, []byte, error) {
```
Comment (yhua123) [inline api/harborapi_test.go]: shouldn't directly return any id(int), better return HEAD structure and get the id in separate function.

File: api/harborapi_test.go [hunk 2]
```diff
@@ -77,19 +80,28 @@ func init() {
...
+		return id, 400, nil, err
 	}
 	if len(authInfo) > 0 {
 		req.SetBasicAuth(authInfo[0].Name, authInfo[0].Passwd)
 	}
 	w := httptest.NewRecorder()
 	beego.BeeApp.Handlers.ServeHTTP(w, req)
 	body, err := ioutil.ReadAll(w.Body)
-	return w.Code, body, err
+	location := w.HeaderMap["Location"]
```
Comment (yhua123) [inline api/harborapi_test.go]: better in separate function

File: api/harborapi_test.go [hunk 3]
```diff
@@ -137,7 +149,49 @@ func (a api) ProjectsPost(prjUsr usrInfo, project apilib.ProjectReq) (int, error
...
+	return userID, httpStatusCode, err
+}
+
+//User Delete
+func (a api) UserDelete(userID int, authInfo ...usrInfo) (int, error) {
+
+	_sling := sling.New().Delete(a.basePath)
+
+	// create path and map variables
+	path := "/api/users/" + fmt.Sprintf("%v", userID)
```
Comment (yhua123) [inline api/harborapi_test.go]: already known id is int format, should use "%d"

File: api/harborapi_test.go [hunk 4]
```diff
@@ -137,7 +149,49 @@ func (a api) ProjectsPost(prjUsr usrInfo, project apilib.ProjectReq) (int, error
...
 	// body params
 	_sling = _sling.BodyJSON(project)
-	httpStatusCode, _, err := request(_sling, jsonAcceptHeader, prjUsr)
+	_, httpStatusCode, _, err := request(_sling, jsonAcceptHeader, prjUsr)
+	return httpStatusCode, err
+
+}
+
+//User Add
+func (a api) UserAdd(user apilib.User, authInfo ...usrInfo) (int, int, error) {
```
Comment (yhua123) [inline api/harborapi_test.go]: //Creates a new user account.
//func (a api) UsersPost (user User) (error) {
//}
please follow standard function name, check all your function name. **include log and hot repo**
...
//Creates a new user account.
//func (a HarborApi) UsersPost (user User) (error) {
//}

//Mark a registered user as be removed.
//func (a HarborApi) UsersUserIdDelete (userId int32) (error) {
//}

//Update a registered user to change to be an administrator of Harbor.
//func (a HarborApi) UsersUserIdPut (userId int32) (error) {
//}

File: api/harborapi_test.go [hunk 5]
```diff
@@ -137,7 +149,49 @@ func (a api) ProjectsPost(prjUsr usrInfo, project apilib.ProjectReq) (int, error
...
+	if len(authInfo) > 0 {
+		userID, httpStatusCode, _, err = request(_sling, jsonAcceptHeader, authInfo[0])
+	} else {
+		userID, httpStatusCode, _, err = request(_sling, jsonAcceptHeader)
+	}
+	return userID, httpStatusCode, err
+}
+
+//User Delete
+func (a api) UserDelete(userID int, authInfo ...usrInfo) (int, error) {
```
Comment (yhua123) [inline api/harborapi_test.go]: wrong function name

File: api/harborapi_test.go [hunk 6]
```diff
@@ -137,7 +149,49 @@ func (a api) ProjectsPost(prjUsr usrInfo, project apilib.ProjectReq) (int, error
...
+	// body params
+	_sling = _sling.BodyJSON(user)
+	var httpStatusCode, userID int
+	var err error
+	if len(authInfo) > 0 {
+		userID, httpStatusCode, _, err = request(_sling, jsonAcceptHeader, authInfo[0])
+	} else {
+		userID, httpStatusCode, _, err = request(_sling, jsonAcceptHeader)
+	}
+	return userID, httpStatusCode, err
```
Comment (yhua123) [inline api/harborapi_test.go]: why return userID?
shouldn't get userID form this function.
better get id from API (get /users)

File: api/harborapi_test.go [hunk 7]
```diff
@@ -137,7 +149,49 @@ func (a api) ProjectsPost(prjUsr usrInfo, project apilib.ProjectReq) (int, error
...
+
+	_sling := sling.New().Delete(a.basePath)
+
+	// create path and map variables
+	path := "/api/users/" + fmt.Sprintf("%v", userID)
+	_sling = _sling.Path(path)
+
+	var httpStatusCode int
+	var err error
+	if len(authInfo) > 0 {
```
Comment (yhua123) [inline api/harborapi_test.go]: why use if here? not better why.
please read project delete code.
---------------

func (a api) ProjectsDelete(prjUsr usrInfo, projectID string) (int, error) {
	_sling := sling.New().Delete(a.basePath)

	//create api path
	path := "api/projects/" + projectID
	_sling = _sling.Path(path)
	_, httpStatusCode, _, err := request(_sling, jsonAcceptHeader, prjUsr)
	return httpStatusCode, err
}
--------

File: api/user_test.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	_, code, err = apiTest.UserAdd(user, *admin)
+	if err != nil {
+		t.Error("Error occured while add a user", err.Error())
+		t.Log(err)
+	} else {
+		assert.Equal(409, code, "Add user status should be 409")
+	}
+}
+
+func TestUserDelete(t *testing.T) {
```
Comment (yhua123) [inline api/user_test.go]: add cases
Handle the error branch.

File: api/harborapi_test.go (hunk 7 above)
Comment (wemeya) [inline api/harborapi_test.go]: There should be a case without auth which I will add later.

File: api/harborapi_test.go (hunk 3 above)
Comment (wemeya) [inline api/harborapi_test.go]: I need pass the user id to the request path, it must be a string in the path.

File: api/harborapi_test.go (hunk 5 above)
Comment (wemeya) [inline api/harborapi_test.go]: what's wrong?

File: api/harborapi_test.go (hunk 6 above)
Comment (wemeya) [inline api/harborapi_test.go]: I can get user id through get function, but xiaoxiao need to call the get function again to get user id to delete after calling the user add function. So I think return user id is better.

File: api/harborapi_test.go (hunk 5 above)
Comment (yhua123) [inline api/harborapi_test.go]: //Mark a registered user as be removed.
//func (a HarborApi) UsersUserIdDelete (userId int32) (error) {
//}

File: api/harborapi_test.go (hunk 6 above)
Comment (yhua123) [inline api/harborapi_test.go]: You should decoupling your code with Xiaoxiao.
Everything she wanted need call from api, shouldn't from you.

File: api/harborapi_test.go [hunk 9]
```diff
@@ -740,22 +740,112 @@ func (a api) DeletePolicyByID(authInfo usrInfo, policyID string) (int, error) {
 //func (a HarborApi) RepositoriesTagsGet (repoName string) (error) {
 //}
 
+
 //Get registered users of Harbor.
-//func (a HarborApi) UsersGet (userName string) ([]User, error) {
-//}
+func (a api) UsersGet(userName string, authInfo usrInfo, userID ...int) (int, apilib.User, error) {
```
Comment (yhua123) [inline api/harborapi_test.go]: please follow project API test code

**Pull request merged**